source/extensions/filters/http/cache/hazelcast_http_cache/hazelcast_http_cache_impl.cc

```diff
@@ -0,0 +1,267 @@
+  cache_info.supports_range_requests_ = true;
+  return cache_info;
+}
+
+void HazelcastHttpCache::putHeader(const uint64_t key, const HazelcastHeaderEntry& entry) {
+  getHeaderMap().set(mapKey(key), entry);
+}
+
+void HazelcastHttpCache::putBody(const uint64_t key, const uint64_t order,
+                                 const HazelcastBodyEntry& entry) {
+  getBodyMap().set(orderedMapKey(key, order), entry);
+}
+
+HazelcastHeaderPtr HazelcastHttpCache::getHeader(const uint64_t key) {
+  return getHeaderMap().get(mapKey(key));
+}
+
+HazelcastBodyPtr HazelcastHttpCache::getBody(const uint64_t key, const uint64_t order) {
+  return getBodyMap().get(orderedMapKey(key, order));
+}
+
+void HazelcastHttpCache::onMissingBody(uint64_t key, int32_t version, uint64_t body_size) {
+  try {
+    if (!tryLock(key)) {
+      // Let lock owner context to recover it.
+      return;
+    }
+    auto header = getHeader(key);
+    if (header && header->version() != version) {
+      // The missed body does not belong to the looked up header. Probably eviction and then
+      // insertion for the header has happened in the meantime. Since new insertion will
+      // override the existing bodies, ignore the cleanup and let orphan bodies (belong to
+      // evicted header, not overridden) be evicted by TTL as well.
+      unlock(key);
+      return;
+    }
+    int body_count = body_size / body_partition_size_;
+    while (body_count >= 0) {
+      getBodyMap().removeAsync(orderedMapKey(key, body_count--));
+    }
+    getHeaderMap().remove(mapKey(key));
+    unlock(key);
+  } catch (HazelcastClientOfflineException e) {
+    // see DividedInsertContext#insertHeader() for left over locks on a connection failure.
+    ENVOY_LOG(warn, "Hazelcast Connection is offline!");
+  } catch (OperationTimeoutException e) {
+    ENVOY_LOG(warn, "Clean up for missing body has timed out.");
+  } catch (...) {
+    ENVOY_LOG(warn, "Clean up for missing body has failed.");
+  }
+}
+
+void HazelcastHttpCache::onVersionMismatch(uint64_t key, int32_t version, uint64_t body_size) {
+  onMissingBody(key, version, body_size);
+}
+
+void HazelcastHttpCache::putResponseIfAbsent(const uint64_t key,
+                                             const HazelcastResponseEntry& entry) {
+  getResponseMap().putIfAbsent(mapKey(key), entry);
+}
+
+HazelcastResponsePtr HazelcastHttpCache::getResponse(const uint64_t key) {
+  return getResponseMap().get(mapKey(key));
+}
+
+bool HazelcastHttpCache::tryLock(const uint64_t key) {
+  // Internal lock mechanism of Hazelcast specific to map and key pair is
+  // used to make exactly one lookup context responsible for insertions and
+  // secure consistency during updateHeaders(). These locks prevent possible
+  // race for multiple cache filters from multiple proxies when they connect
+  // to the same Hazelcast cluster.
+  // The locks used here are re-entrant. A locked key can be acquired by
+  // the same thread again and again based on its pid.
```

Comment:
As I examined the threading model of Envoy, I concluded that the same thread does lookup and then insertion (if requested) for the filter. Is that actually true? In other words, can a particular thread handle a lookup `L1` and then another  lookup `L2` before `L1` (and its insertion if requested) is over?

Reply:
Yes, the same thread does lookup and then insertion (if requested) for the filter. However, that same thread will do any number of other things in between, such as doing lookups and/or insertions for other requests.

Reply:
This is fine as I moved the map's key locking logic to insert contexts before you reviewed.